### PR TITLE
rustdoc: revert deref recur to resume inclusion of impl ExtTrait<Local> for ExtType

### PR DESCRIPTION
**Pull request by pnkfelix**
As discussed here: https://github.com/rust-lang/rust/issues/82465#issuecomment-829290384, Revert PR #80653 to resolve issue #82465.

Issue #82465 was we had stopped including certain trait implementations, namely implementations on an imported type of an imported trait *instantiated on a local type*. That bug was injected by PR #80653.

Reverting #80653 means we don't list all the methods that you have accessible via recursively applying `Deref`.

[Discussion in last week's rustc triage meeting](https://zulip-archive.rust-lang.org/238009tcompilermeetings/19557weekly2021042954818.html#236680594) led us to conclude that the bug was worse than the enhancement, and there was not an obvious fix for the bug itself. So for the short term we  remove the enhancement, while in the long term we will work on figuring out a way to have our imported trait implementation cake and eat it too.

### CODE REVIEW

File: src/test/rustdoc/issue-82465-asref-for-and-of-local.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+        todo!()
+    }
+}
+
+// @has - '//code' 'impl AsRef<Local> for str'
+impl AsRef<Local> for str {
+    fn as_ref(&self) -> &Local {
+        todo!()
+    }
+}
```
Comment (jyn514) [inline src/test/rustdoc/issue-82465-asref-for-and-of-local.rs:16]: Thanks for this test case :)

**Review by jyn514 — APPROVED** [review]
@pnkfelix can you also revert this on nightly? I don't feel comfortable having different behavior on beta, the underlying issue won't be fixed by the next release.

cc @GuillaumeGomez @jryans

**Comment by jyn514** [comment]
Oh, this is for nightly - I guess it needs to be backported to beta.

**Comment by jyn514** [comment]
Err I guess CI is failing - do you mind fixing it real quick?

@bors r-

File: src/librustdoc/html/render/mod.rs [hunk 2]
```diff
@@ -1952,11 +1943,14 @@ fn sidebar_deref_methods(cx: &Context<'_>, out: &mut Buffer, impl_: &Impl, v: &V
             .iter()
             .filter(|i| i.inner_impl().trait_.is_some())
             .any(|i| i.inner_impl().trait_.def_id_full(c) == c.deref_mut_trait_did);
-        let inner_impl = target
-            .def_id_full(c)
+        let inner_impl = target.def_id_full(c);
+        debug!("target.def_id_full: {:?}", inner_impl);
```
Comment (pnkfelix) [inline src/librustdoc/html/render/mod.rs:1947]: (sigh, I wonder if this extra debugging lines were part of why this didn't backport cleanly...)

**Comment by pnkfelix** [comment]
I'm unilaterally beta-accepting this. Even though its somewhat hand-crafted, its almost entirely based on eyeball review of PR #80653 , so I think it is in the spirit of the revert that the T-compiler team agreed upon in last week's triage meeting.

**Comment by pnkfelix** [comment]
the beta-targeted version is PR #84868

**Comment by pnkfelix** [comment]
(looking at CI failure now)

**Comment by pnkfelix** [comment]
oh. good old tidy. (I usually automatically run tidy, but I think there's some reason that I cannot run tidy when I'm targetting beta channel, at least when I do it and actually choose `channel=beta` in config.toml)

**Comment by pnkfelix** [comment]
(this will give me a chance to remove the `debug!` lines I left in by accident. 😄 )

**Comment by jryans** [comment]
Thanks all for managing this, hopefully we'll be able to revive the defer recursion without this class of bugs in the future.

**Comment by jyn514** [comment]
@jryans you did a great job in the original PR by the way - we only had to revert this because rustdoc has long-standing bugs with resolve: https://github.com/rust-lang/rust/issues/83761

**Comment by Mark-Simulacrum** [comment]
Note: this is targeting 1.53 for the beta-nomination; it's bundled into 1.52 via #84864.

**Comment by jyn514** [comment]
This fix is incorrect, see https://github.com/rust-lang/rust/pull/84867.

**Comment by jyn514** [comment]
I won't have time to review this.

r? @GuillaumeGomez

**Comment by jyn514** [comment]
oh err I guess @CraftSpider was assigned originally, are you comfortable reviewing this?

**Comment by CraftSpider** [comment]
Iffy, I'll at least do a review, but I'm fine with Guillame taking it if they're okay with it.

**Comment by GuillaumeGomez** [comment]
I don't mind reviewing it. Let's do this: @CraftSpider will review it and then I'll review as well later on. Like that you will get used to reviews. ;)

**Comment by JohnCSimon** [comment]
triage:
@pnkfelix - there is still a merge conflict. Can you please resolve it?

**Comment by jyn514** [comment]
Note that this has already been bacjported to 1.53 (currently beta): https://rust-lang.zulipchat.com/#narrow/stream/241545-t-release/topic/how%20to%20track%20ETA%20when%20nightly%20will%20be%20cut%20to%20beta/near/242135597

**Comment by pnkfelix** [comment]
> Note that this has already been bacjported to 1.53 (currently beta): https://rust-lang.zulipchat.com/#narrow/stream/241545-t-release/topic/how%20to%20track%20ETA%20when%20nightly%20will%20be%20cut%20to%20beta/near/242135597

This is a semi-interesting corner case in our labels for beta-nominated/beta-accepted.

In particular, the fact that this has landed in 1.53-beta might lead one to conclude that the beta-nominated label should be removed.

However, the fact that it *hasn't* landed in 1.54-nightly to me implies that the labels should stay as they are, to try to ensure that if 1.54-beta is cut, that we remember to *then* backport this from 1.55-nightly to 1.54-beta.

(However, I personally am hoping the fix I am currently locally testing will work out and that I'll be able to get this landed before 1.54-beta is cut, at which point it will have landed everywhere we want, and we can then remove the appropriate backport labels ...)

**Comment by CraftSpider** [comment]
Sorry, meant to review this when you fixed the merge conflict, but it slipped through. Thanks joshua for handling this

**Pull request merged**